### PR TITLE
JWE access token implementation

### PR DESCRIPTION
**Pull request by RishiDiwanTT**
## Description

Token creation endpoint in implemented
Token decode and key generation has been implemented in the auth class
Implemented JWE key rotation scripts and cron
Refactored api.authenticator by moving out the base classes into their own files
since the file size had gotten to >2000 lines and kept growing

This mechanism has only been implemented for the basic authentication type authorization because the SAML authorization is already local.

The provider has an authentication document component

<details>
<summary>Authentication Document</summary>

```
{
    "description": "Library Barcode + Token",
    "type": "http://librarysimplified.org/authtype/basic-token",
    "inputs": {
        "login": {
            "keyboard": "default"
        },
        "password": {
            "keyboard": "default"
        }
    },
    "links": [
        {
            "rel": "authenticate",
            "href": "http://localhost:6500/localtest/patrons/me/token/"
        }
    ],
    "labels": {
        "login": "Barcode",
        "password": "Pin"
    }
}
```
</details>

## Motivation and Context
We needed to implement token based authentication for our API endpoints, so that we do not have to always validate a barcode / pin against the client ILS.

[Notion](https://www.notion.so/lyrasis/Implement-token-based-authentication-f601474812284d449ba5f61ca0b9401e)

## How Has This Been Tested?
Local unit tests have been updated and run.
One can request an access token via `GET /<library>/patrons/me/token` with Basic Auth.
The `Bearer` access token can be used with any authenticated API like `/patrons/me` or `/loans`.

## Checklist

- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: api/admin/model/dashboard_statistics.py
Comment (RishiDiwanTT) [inline api/admin/model/dashboard_statistics.py:7]: Moved this to an accessible location

File: api/authentication/access_token.py
Comment (RishiDiwanTT) [inline api/authentication/access_token.py:40]: This class only provides the JWE token, it does not do authentication

File: api/authentication/base.py
Comment (RishiDiwanTT) [inline api/authentication/base.py:1]: Just refactoring changes

File: api/authentication/basic.py
Comment (RishiDiwanTT) [inline api/authentication/basic.py:1]: Refactoring changes

File: api/authentication/patron_authentication_provider.py
Comment (RishiDiwanTT) [inline api/authentication/patron_authentication_provider.py:13]: This is an imaginary flow type, if there are semantics to this we should change it.

File: core/util/authentication_for_opds.py
Comment (RishiDiwanTT) [inline core/util/authentication_for_opds.py:81]: This allows us to skip flows in the authentication document

File: api/authentication/access_token.py
Comment (io7m) [inline api/authentication/access_token.py:115]: Just wanted to confirm here that we're not using a predictable `iv`, given that we're using AES 128 CBC.

File: core/util/authentication_for_opds.py
Comment (io7m) [reply core/util/authentication_for_opds.py:81]: Why do we need to skip flows?

File: api/authentication/patron_authentication_provider.py
Comment (io7m) [reply api/authentication/patron_authentication_provider.py:15]: I think the mobile clients will inspect this value to decide whether a server supports token auth, right?

If so, yeah, we should decide if this is the constant we want. Extremely hard to change afterwards.

File: core/util/flask_util.py
Comment (RishiDiwanTT) [reply core/util/flask_util.py:189]: Hmm.. I didn't write this method, just moved it here, I can add tests for it though

File: bin/rotate_patron_auth_key
Comment (RishiDiwanTT) [reply bin/rotate_patron_auth_key:2]: No SIR! I did not copy from a another file!!!

File: core/util/authentication_for_opds.py
Comment (io7m) [reply core/util/authentication_for_opds.py:81]: I'm kind of unsure about this. The clients used basic authentication to retrieve a token, that part I understand. The problem is that (given that I know a bit about how the Android client is implemented), they're going to need to understand that this "basic authentication" behaves differently to the other kind of "basic authentication" they'll see elsewhere, and that they'll need to store and use a token repeatedly. I feel like this information _should_ be in the authentication document.

File: core/util/flask_util.py
Comment (io7m) [reply core/util/flask_util.py:189]: Oh, I didn't realize that.

File: bin/rotate_patron_auth_key
Comment (RishiDiwanTT) [reply bin/rotate_patron_auth_key]: I admit guilt, and have changed my ways.

File: core/util/flask_util.py
Comment (io7m) [reply core/util/flask_util.py:189]: Thanks for adding tests! This is a public service. :laughing:

File: core/util/authentication_for_opds.py
Comment (RishiDiwanTT) [reply core/util/authentication_for_opds.py:81]: I don't think there is any documented protocol to define that in the authentication document. I think the prevailing thought here was that this would be written in directly into the code of the Basic Auth implementations and not controlled via the authentication document.

Comment (RishiDiwanTT) [reply core/util/authentication_for_opds.py:81]: Eg.
```
"authentication": [
    {
        "description": "Library Barcode",
        "labels": {
            "login": "Barcode",
            "password": "PIN"
        },
        "inputs": {
            "login": {
                "keyboard": "Default"
            },
            "password": {
                "keyboard": "Default"
            }
        },
        "type": "http://opds-spec.org/auth/basic",
        "secondary": {
           "type": "http://opds-spec.org/auth/bearer",
           "href": ".../patrons/me/token",
        }
    }
],
```

Comment (io7m) [reply core/util/authentication_for_opds.py:81]: I feel like it should be treated as an entirely separate authentication method, with a different `type`.

Doing so unambiguously signals to the client that they can use a different login workflow to use tokens (and by putting the token option in the authentication document ahead of plain basic authentication, it signals to them that they _should_ use tokens in preference to plain basic authentication). The clients already have separate code paths for basic auth, OAUTH, and SAML, so adding another authentication method shouldn't be a problem.

It also allows us to turn off plain basic authentication eventually, when we've determined that no clients have used it for _insert time period here_.

Comment (io7m) [reply core/util/authentication_for_opds.py:81]: As an aside, the Android client is using sealed types for the credentials, so if @nunommts adds a new `PalaceBearerToken` credentials type, he'll get nice errors indicating all of the places in the codebase that now need updating. :slightly_smiling_face:

**Comment by RishiDiwanTT** [comment]
@jonathangreen @io7m Authentication Document implemented

**Comment by jonathangreen** [comment]
To make the conflict on this PR with #1063 less nightmarish, would it make sense to put the refactor of `api.authenticator` into its own module, into its own separate PR?

**Comment by RishiDiwanTT** [comment]
> To make the conflict on this PR with #1063 less nightmarish, would it make sense to put the refactor of `api.authenticator` into its own module, into its own separate PR?

I don't think it'll be easy to revert the refactor, since the JWE implementation and the refactor are in a single commit. Stupid, I know.
It's probably the same amount of work at this point in time, so I'll keep this updated as we go along.

File: api/authentication/access_token.py
Comment (RishiDiwanTT) [inline api/authentication/access_token.py:88]: @jonathangreen The fix was to have a completely new dict object as opposed to changing the current object

Comment (RishiDiwanTT) [reply api/authentication/access_token.py:88]: The solution came while I was asking phind.com for random solutions :sweat_smile: . Time to be replaced :robot:

**Review by jonathangreen — COMMENTED** [review]
Started to review this today. I'll keep going with it tomorrow. One thing that is making this a bit harder to review is the refactor mixed in. It would be nice to make that its own PR, so its easier to read the diff here. 

I attempted to move around the same classes, with no changes in this PR https://github.com/ThePalaceProject/circulation/pull/1138, so we could merge the refactor separately. If that PR isn't helpful feel free to close it and make a new branch, or make any changes there.

File: api/authentication/patron_authentication_provider.py
Comment (RishiDiwanTT) [reply api/authentication/patron_authentication_provider.py:84]: Where would we set these values if we are not setting them through the UI? Or do we want to have this provider show up in the `patron_auth_services` ?

File: api/authentication/access_token.py
Comment (RishiDiwanTT) [reply api/authentication/access_token.py:53]: I've created a new `systemkeys` table, the credentials table is geared towards Patrons.
I wanted to use sitewide settings, but is deprecated now with the configurationsettings table. Which brings to mind a different question of how those will be supported?

File: api/authentication/patron_authentication_provider.py
Comment (RishiDiwanTT) [reply api/authentication/patron_authentication_provider.py]: Ok, I've changed the `settings` to inherit from the Basic Auth settings, and use those values in the authentication document. However I see no use of these values in the document since the access token provider should not have a UI for the values anyways, nor should it be presented to the user in any way.
It is a secondary authentication mechanism which does not have the features of the basic auth mechnism.

Comment (RishiDiwanTT) [reply api/authentication/patron_authentication_provider.py]: Based on our slack conversation I have changed the Access token provider to shadow the basic auth provider, from inception to authentication document creation.

**Review by jonathangreen — COMMENTED** [review]
Functionally this looks good, tested and we're generating tokens and can use them. 

Made a few minor comments. My biggest concern is around the class hierarchy structure here. I'm wondering if `PatronAccessTokenAuthenticationProvider` doesn't really make sense to inherit from `AuthenticationProvider`.

File: api/authentication/patron_authentication_provider.py
Comment (RishiDiwanTT) [reply api/authentication/patron_authentication_provider.py:21]: Nope, will remove

File: tests/api/mock_authentication_provider.py
Comment (RishiDiwanTT) [reply tests/api/mock_authentication_provider.py:5]: Looks like a merge issue, will remove it

File: api/authentication/patron_authentication_provider.py
Comment (RishiDiwanTT) [reply api/authentication/patron_authentication_provider.py:1]: I didn't quite like it myself. How about `basic_token.py::BasicTokenAuthenticationProvider`?

Comment (jonathangreen) [reply api/authentication/patron_authentication_provider.py:1]: That sounds reasonable to me.

**Comment by RishiDiwanTT** [comment]
@jonathangreen I think all comments have been addresses

File: api/sirsidynix_authentication_provider.py
Comment (jonathangreen) [reply api/sirsidynix_authentication_provider.py:28]: @RishiDiwanTT I think this one still needs to be addressed.

**Review by jonathangreen — APPROVED** [review]
I think this looks good. 

Would be nice to address these two comments before merging though: 
- https://github.com/ThePalaceProject/circulation/pull/1076#discussion_r1210306366
- https://github.com/ThePalaceProject/circulation/pull/1076#discussion_r1210300813

**Pull request merged**
